core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -537,14 +533,14 @@ class GroupCoordinator(val brokerId: Int,
           // TODO: cut the socket connection to the client
         }
 
+        // TODO KAFKA-2720: only remove group in the background thread
         if (group.isEmpty) {
-          group.transitionTo(Dead)
+          groupManager.removeGroup(group)
```

Comment:
We have the "Dead" state in which removeGroup should always transit to within the lock, and we always check the state again after we grab the group lock. So I think it should be fine.